{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deliverables:  40 total points\n",
    "\n",
    " - this notebook with inspection intro and results filled out for CoordinateMathSoln  4 points\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as the notebook and 4(or 5) .py files, \n",
    "     \n",
    " Due: Monday January 17, 2022 11:55pm\n",
    "\n",
    "# Lab 2 – Manipulating Data Types (40 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment, and as a final step we will create distinct command line programs from each of the parts.\n",
    "Please save each of the four (or five) python programs as .py files into your LAB02 folder. You can copy the text from the code cell, then paste it into an editor like notepad or textedit. Save that new file with the appropriate name ( eg. seqCleaner.py) into your LAB02 folder. Using terminal or cmd, you can now do a final test on that new program with the command:<br>\n",
    "python3 seqCleaner <br>\n",
    "\n",
    "All together, you will submit:\n",
    " - this notebook, with the inspection sections filled out\n",
    " - seqCleaner.py, \n",
    " - fastqParse,py, \n",
    " - coordinateMathSoln.py, \n",
    " - converter.py, and \n",
    " - (optionally) Triad.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence cleanup\n",
    "In this exercise, you will create a program to “clean up” a sequence of DNA by removing ambiguous bases (denoted by “N”) output from a sequencer. Your task is to create a Python program called seqCleaner that\n",
    " - asks for and collects a sequence of DNA using input()\n",
    " - removes the ambiguous parts of the sequence, outputs the “cleaned” sequence, replacing the ambiguous parts with a count in {}’s.\n",
    "For example, if I enter the sequence of DNA “AaNNNNNNGTC” (without quotes), the program will output:<br>\n",
    "AA{6}GTC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hints: \n",
    " - The input sequence is not guaranteed to be uppercase, but should be interpreted as though it is all uppercase.\n",
    " - Only the letters (A,C,G,T,N) will be included in the input.\n",
    " - Only the DNA characters (A,C,G,T) should remain after the cleanup. \n",
    " - The input may include, at most, one block of 'N' characters. \n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and specific line or block comments that explain the semantics of your implementation.\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs. This should be in the form of a  program level dosctring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a DNA sequence: NNCAT\n",
      "{2}CAT\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Madison Price (maeprice@ucsc.edu)\n",
    "# Group Members: Kiana Imani (kimani) and Queenie Li (qli86)\n",
    "\n",
    "'''\n",
    "Read a DNA string from user input and return a collapsed substring of embedded Ns to: {count}.\n",
    "\n",
    "Example: \n",
    " input: AaNNNNNNGTC\n",
    "output: AA{6}GTC\n",
    "\n",
    "Any lower case letters are converted to uppercase\n",
    "\n",
    "Assumptions: The only letters the user might enter are A, C, T, G (upper or lowercase), and N (uppercase)\n",
    "Design Decisions: I decided to delete the unnecessary class method, length() \n",
    "'''\n",
    "\n",
    "class DNAstring (str):\n",
    "    '''Build a DNAstring object from a string.'''\n",
    "    def purify(self):\n",
    "        ''' Return an upcased version of the string, collapsing a single run of Ns.'''\n",
    "        purified = []                                     # initiate a list for the unambigious letters\n",
    "        count = 0                                         # initiate a count variable\n",
    "        for i in self:                    \n",
    "            if i != 'N':                                  # two possibilities for nonambiguous letters\n",
    "                if count == 0:                            # 1st possibility = no ambiguous letters have been encountered yet\n",
    "                    purified.append(i)                    # append unambiguous letter to the list\n",
    "                if count != 0:                            # 2nd possibility = we have reached an unambiguous letter after a block of ambiguous letters\n",
    "                    purified.append('{'+str(count)+'}')   # append the count of ambiguous letters to the list\n",
    "                    purified.append(i)                    # append unambiguous letter to the list\n",
    "                    count = 0                             # reset count variable so it isn't added to the list more than once\n",
    "            if i == 'N':                                  # if letter is ambigious add it to the count variable\n",
    "                count += 1\n",
    "        if count >= 1 and len(purified) != 0:             # test for case of N's at end of sequence\n",
    "            purified.append('{'+str(count)+'}')\n",
    "        if count >= 1 and len(purified) == 0:             # test for case of only N's\n",
    "             purified.append('{'+str(count)+'}')\n",
    "        purifiedString = ''.join(purified)                # turn DNA sequence as a list into a string\n",
    "        upcasedPurifiedString = purifiedString.upper()    # make sure string is all uppercase\n",
    "        return upcasedPurifiedString                      \n",
    "    \n",
    "def main():\n",
    "    ''' Get user DNA data and clean it up using DNAstring class.'''\n",
    "    data = input('Enter a DNA sequence: ')\n",
    "    thisDNA = DNAstring (data)\n",
    "    pureData = thisDNA.purify()\n",
    "    print (pureData)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence information parsing\n",
    "In this exercise, you will create a program to “parse” sequence name information from a single line of a FASTQ formatted file. Your task is to create a Python script called fastqParse that:\n",
    " - asks for and collects the seqname line of a FASTQ file using input()\n",
    " - parses out each field of the run information from the string and displays each of them on a new line\n",
    "For example, if I enter the FASTQ seqname line:<br> \n",
    "__*@EAS139:136:FC706VJ:2:2104:15343:197393*__\n",
    "then the program will output:<br>\n",
    "Instrument = EAS139<br>\n",
    "Run ID = 136<br>\n",
    "Flow Cell ID = FC706VJ<br>\n",
    "Flow Cell Lane = 2<br>\n",
    "Tile Number = 2104<br>\n",
    "X-coord = 15343<br>\n",
    "Y-coord = 197393\n",
    "### Hints: \n",
    " - The input string is guaranteed to have 7 fields. \n",
    " - The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”.  \n",
    " - A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain documentation/comments\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter seqname line: @EAS139:136:FC706VJ:2:2104:15343:197393\n",
      "Instrument = EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell Lane = 2\n",
      "Tile Number = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Madison Price (maeprice)\n",
    "# Group Members: Kiana Imani (kimani) and Queenie Li (qli86)\n",
    "\n",
    "'''\n",
    "“Parse” sequence name information from a single line of a FASTQ formatted file, where the line is obtained from user input.\n",
    "\n",
    "input: seqname line of a FASTQ file\n",
    "output: parsed out fields of the run information\n",
    "\n",
    "Assumptions: The input string has seven fields (separated by colons) and starts with \"@\"\n",
    "Design Decisions: I used concatenation within print() instead of format methods\n",
    "'''\n",
    "\n",
    "class FastqString (str):\n",
    "    ''' Create a FastqString object from a string.'''\n",
    "    def parse(self):\n",
    "        ''' Parse self into fields that are displayed on separate lines.'''\n",
    "        fields = self[1:]                              # eliminate \"@\"\n",
    "        splitFields = fields.split(\":\")                # split input line around colons and add fields to a list\n",
    "        print(\"Instrument = \"+splitFields[0])          # print each field on a separate line with its own introduction\n",
    "        print(\"Run ID = \"+splitFields[1])\n",
    "        print(\"Flow Cell ID = \"+splitFields[2])\n",
    "        print(\"Flow Cell Lane = \"+splitFields[3])\n",
    "        print(\"Tile Number = \"+splitFields[4])\n",
    "        print(\"X-coord = \"+splitFields[5])\n",
    "        print(\"Y-coord = \"+splitFields[6])\n",
    "        \n",
    "def main():\n",
    "    '''Get user seqname line data and format it using FastqString class.'''\n",
    "    seqnameLineData = input(\"Enter seqname line: \")\n",
    "    fastqStringObject = FastqString (seqnameLineData)\n",
    "    fastqStringObject.parse()\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein coordinates\n",
    "In this exercise, you will create a program that takes three sets of atomic coordinates, all provided on a **single line**. The program then calculates the bond lengths and angles. For this program, you can start with the Triad class (provided). Your task is to create a Python program called coordinateMathSoln that:\n",
    " - asks for and collects three sets of coordinates using input(), only use 1 line for this data !!\n",
    " - outputs the N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits (see below)\n",
    " \n",
    "For example, if I enter the following coordinates (notice.. they are all on one line !!!) :<br>\n",
    "\n",
    "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465) <br>\n",
    "\n",
    "then the program will output the following three lines:<br>\n",
    "N-C bond length = 1.33<br>\n",
    "N-Ca bond length = 1.46<br>\n",
    "C-N-Ca bond angle = 124.0<br> \n",
    "(Note: make sure that the angle returned is in **degrees** !!)\n",
    "### Hints: \n",
    " - Each coordinate will contain only 3 numeric values. \n",
    " - Bond lengths are the distance between two points, which for points P and Q in 3-space,\n",
    "$(P_x , P_y , P_z )$ and $(Q_x , Q_y , Q_z )$ respectively, the distance between them is:\n",
    "\n",
    "\n",
    "$$\n",
    "\\begin {eqnarray}\n",
    "     \\|PQ\\| &=&\\sqrt{(P_x-Q_x)^2 + (P_y-Q_y)^2 + (P_z-Q_z)^2)} \\\\\n",
    "            &=&\\sqrt{\\sum^{i \\in x,y,z}{(P_i-Q_i)^2}} \\\\\n",
    "\\end{eqnarray}\n",
    "$$\n",
    "\n",
    " - Bond angles can be calculated from the dot product.\n",
    " \n",
    "Let’s say that we have three points in space labeled P, Q and R. We are interested in the angle at point Q that is made from line segments QP and QR.\n",
    "The dot product tells us (for standard vectors P and R) that:\n",
    "$$P \\cdot R=\\|P\\| \\|R\\| \\cos{\\theta}$$\n",
    "in this notation,$\\|P\\|$ refers to the length of vector P as a standard vector (assumed to begin at the origin (0,0,0) ).\n",
    "We can then see that the angle between vectors P and R can be found by:\n",
    "$$\\cos{\\theta}=\\frac{P \\cdot R} {\\|P\\| \\|R\\|}$$\n",
    "\n",
    "We can calculate the dot product using the sum of products of the vector components:\n",
    "$$P \\cdot R = \\sum^{i \\in x,y,z}{P_i R_i} $$\n",
    "\n",
    "Now, to find vector P in standard form, we need to remember that QP starts at Q, so we need to place the origin at Q and find out where P is in that new space. We do that by subtracting the components of Q from P.\n",
    "Putting all of this together, we get:\n",
    "$$\\theta=\\cos^{-1}{\\frac{\\sum^{i \\in x,y,z}{(P_i - Q_i) (R_i - Q_i) }} {\\|QP\\| \\|QR\\|}}$$\n",
    "Remember, $\\theta$ is in radians.\n",
    "\n",
    "Below I have given you a class (Triad) with methods to calculate dot products (dot), dot products of translated vectors (ndot), distances (dPQ, dPR, dPR) and angles in radians (angleP, angleQ and angleR) for each of the three points in a Triad object.\n",
    "A reasonable solution for this exercise involves around 12 lines of additional code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Introduction\n",
    "\n",
    "For this wweeks inspection please focus on coordinateMathSoln. COnsider the design choices that you made, how you are dealing with the single line of input data, how you are doing the data conversions, and how you convert to degrees.\n",
    "\n",
    "I dealt with the single line of input data by replacing one type of parenthesis with the other, and split the input string around that type of parenthesis. I add these split strings to a list, and define each set of atomic coordinates by referencing the appropriate element of the list and using the eval() method to convert each list element (which are strings) to tuples. I then create a Triad object using three tuples, and apply methods of the Triad class. I convert from radians to degrees using a simply conversion factor within a print statement.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter three sets of atomic coordinates: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-N-Ca bond angle = 124.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Madison Price (maeprice)\n",
    "# Group Members: Kiana Imani (kimani) and Queenie Li (qli86)\n",
    "\n",
    "'''\n",
    "Take three sets of atomic coordinates and calculate bond lengths and bond angles.\n",
    "\n",
    "input: three sets of atomic coordinates\n",
    "output: two bond lengths and one bond angle\n",
    "\n",
    "Assumptions: The input line contains three sets of coordinates, each of which contains 3 numeric values.\n",
    "Design Decisions: I deal with the single line of input by splitting it around any parentheses to yield individual strings with atomic coordinates.\n",
    "                  I used eval() to convert the strings to tuples to be used by the Triad class.\n",
    "                  I convert from radians to degrees using a single unit conversion.\n",
    "'''\n",
    "\n",
    "import math\n",
    "\n",
    "\n",
    "class Triad:\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    "\n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    "\n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, p, q, r):\n",
    "        \"\"\" Construct a Triad.\n",
    "\n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ).\n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "\n",
    "    # private helper methods\n",
    "    def d2(self, a, b):  # calculate squared distance of point a to b\n",
    "        return float(sum((ia - ib) * (ia - ib) for ia, ib in zip(a, b)))\n",
    "\n",
    "    def dot(self, a, b):  # dotProd of standard vectors a,b\n",
    "        return float(sum(ia * ib for ia, ib in zip(a, b)))\n",
    "\n",
    "    def ndot(self, a, b, c):  # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia - ib) * (ic - ib) for ia, ib, ic in zip(a, b, c)))\n",
    "\n",
    "    # calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ(self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p, self.q))\n",
    "\n",
    "    def dPR(self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p, self.r))\n",
    "\n",
    "    def dQR(self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q, self.r))\n",
    "\n",
    "    def angleP(self):\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(\n",
    "            self.ndot(self.q, self.p, self.r) / math.sqrt(self.d2(self.q, self.p) * self.d2(self.r, self.p)))\n",
    "\n",
    "    def angleQ(self):\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(\n",
    "            self.ndot(self.p, self.q, self.r) / math.sqrt(self.d2(self.p, self.q) * self.d2(self.r, self.q)))\n",
    "\n",
    "    def angleR(self):\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(\n",
    "            self.ndot(self.p, self.r, self.q) / math.sqrt(self.d2(self.p, self.r) * self.d2(self.q, self.r)))\n",
    "\n",
    "\n",
    "\n",
    "def main():\n",
    "    ''' Get atomic coordinates from user and use Triad class to calculate bond lengths and angles. '''\n",
    "    coords = input(\"Enter three sets of atomic coordinates: \")\n",
    "    atom1 = \"C\"\n",
    "    atom2 = \"N\"\n",
    "    atom3 = \"Ca\"\n",
    "\n",
    "    modifiedCoords = coords.replace(\")\", \"(\")          # make sure string contains only one type of paranthesis\n",
    "    splitCoords = modifiedCoords.split(\"(\")            # split string around that type of paranthesis, and add components to a list\n",
    "    atom1Coords = eval(splitCoords[1])                 # turn elements of the list into tuples and assign approrpriate names\n",
    "    atom2Coords = eval(splitCoords[3])\n",
    "    atom3Coords = eval(splitCoords[5])\n",
    "    triad = Triad(atom1Coords, atom2Coords, atom3Coords)   # create a Traid object with the tuples (coords)\n",
    "\n",
    "    print(atom2 + \"-\" + atom1 + \" bond length = \" + \"{:.2f}\".format(triad.dPQ()))                      # print bond lengths and angle using Triad methods\n",
    "    print(atom2 + \"-\" + atom3 + \" bond length = \" + \"{:.2f}\".format(triad.dQR()))\n",
    "    print(atom1 + \"-\" + atom2 + \"-\" + atom3 + \" bond angle = \" + \"{:.1f}\".format(57.3*triad.angleQ())) # convert radians to degrees using a concersion factor\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Results \n",
    "(Group, Findings and your Responses to those findings)\n",
    "\n",
    "My group includes Kiana Imani, Elizabeth Beer, and Queenie Li. They liked my design choice to delete the len() class method in SeqCleaner, as they agred that it seemed unecessary and thought that its deletion made my code look 'cleaner'. They liked my inclusion of 'assumptions' in class docstrings, and agreed that my use of comments made understanding my programs easier. They reminded me to uppercase the output of my converter function, so I made use of the upper() function. They also caught that my SeqCleaner program didn't satisfy the condition of a sequence of only N's, so I added a conditional statement to fix that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra credit (5 points): Rewrite the Triad class.  \n",
    "For extra-credit, provide a direct replacement for the Triad class. The external methods that calculate angles, distances, and points (tuples) p,q and r must be maintained such that either version of the Triad class can be used.\n",
    "\n",
    "You could use the cosine law to calculate angles instead of the dot product.  You might make use of the numpy module. You might recode each of the methods to avoid using zip. You might consider using list iterations.\n",
    "Your Triad replacement must reimplement all of Triad public function, without using zip and without being a trivial rewrite. Your implementation need not be as compact as the current implementation, and it needs to be correct and fully documented to receive full credit. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Triad rewrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "\n",
    "class Triad:\n",
    "    def __init__(self, p, q, r):\n",
    "        \"\"\" Construct a Triad.\n",
    "\n",
    "        Example object construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ).\n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "\n",
    "    def dPQ(self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(((self.p[0] - self.q[0]) ** 2) + ((self.p[1] - self.q[1]) ** 2) + (\n",
    "                    (self.p[2] - self.q[2]) ** 2))                                                  # pull coordinate values directly from tuples\n",
    "\n",
    "    def dPR(self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(((self.p[0] - self.r[0]) ** 2) + ((self.p[1] - self.r[1]) ** 2) + (\n",
    "                    (self.p[2] - self.r[2]) ** 2))                                                  # pull coordinate values directly from tuples\n",
    "\n",
    "    def dQR(self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(((self.q[0] - self.r[0]) ** 2) + ((self.q[1] - self.r[1]) ** 2) + (\n",
    "                    (self.q[2] - self.r[2]) ** 2))                                                  # pull coordinate values directly from tuples\n",
    "\n",
    "    def angleP(self):\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians) using the law of cosines. \"\"\"     # use law of cosines\n",
    "        return math.acos((((self.dPQ() ** 2) + (self.dPR()) ** 2) - (self.dQR() ** 2)) / (2 * self.dPQ() * self.dPR()))\n",
    "\n",
    "    def angleQ(self):\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians) using the law of cosines. \"\"\"      # use law of cosines\n",
    "        return math.acos((((self.dPQ() ** 2) + (self.dQR()) ** 2) - (self.dPR() ** 2)) / (2 * self.dPQ() * self.dQR()))\n",
    "\n",
    "    def angleR(self):\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians) using the law of cosines. \"\"\"      # use law of cosines\n",
    "        return math.acos((((self.dPR() ** 2) + (self.dQR()) ** 2) - (self.dPQ() ** 2)) / (2 * self.dPR() * self.dQR()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon tables and amino acid letter converters\n",
    "In this exercise, you will create a program that uses mappings to convert sequence information between different amino acid representations.  This includes the 3-letter codon code (RNA and DNA), the one letter amino acid code and the 3-letter amino acid code.\n",
    "The program will use different dictionaries that represent :\n",
    "codon tables – one for DNA and the other for RNA, and\n",
    "amino acid letter representation converters. \n",
    "\n",
    "Your task is to create a Python program called converter that\n",
    "asks for and collects a single input string using input()\n",
    "parses the string, looks up the information in the appropriate dictionary, and outputs the correct conversion\n",
    "For example:<br>\n",
    "\n",
    "if I enter “ATG” (without quotes), then the program will output:<br>\n",
    "ATG = MET<br>\n",
    " \n",
    "if I enter “UAG” (without quotes), then the program will output:<br>\n",
    "UAG = ---<br>\n",
    "if I enter “E” (without quotes), then the program will output:<br>\n",
    "E = GLU<br>\n",
    "if I enter “Asp” (without quotes), then the program will output:<br>\n",
    "ASP = D<br>\n",
    "### Hints: \n",
    " - The program might not get a valid codon.  In that case, it should output 'unknown'.  You can use the dictionary method 'get' and include a default_value to handle the 'unknown' case.  See Model p. 72 for an example.\n",
    " \n",
    "To get full credit on this assignment, your program needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter an amino acid representation: ASP\n",
      "ASP = D\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Madison Price (maeprice)\n",
    "# Group Members: Kiana Imani (kimani) and Queenie Li (qli86)\n",
    "\n",
    "'''\n",
    "Use mappings to convert sequence information between different amino acid representations.\n",
    "\n",
    "Inputs: three letter codon, one letter amino acid code, or three letter amino acid code\n",
    "Outputs: one letter amino acid code (for three letter aa code), three letter aa code (for three letter codon)\n",
    "\n",
    "Assumptions: No assumptions made\n",
    "Design Decisions: I used the get() method to search all four dictionaries at once, and included a default value of \"unknown\" \n",
    "'''\n",
    "short_AA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }\n",
    "\n",
    "long_AA = {value:key for key,value in short_AA.items()}\n",
    "\n",
    "rnaCodonTable = {\n",
    "# Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "def main():\n",
    "    ''' Collect a single input string using input(), parse the string, look up the information in \n",
    "    the appropriate dictionary, and outputs the correct conversion.'''\n",
    "    data = input(\"Enter an amino acid representation: \")                                                                 # ask user for sequence information\n",
    "    upperdata = data.upper()\n",
    "    result = short_AA.get(upperdata) or long_AA.get(upperdata) or rnaCodonTable.get(upperdata) or dnaCodonTable.get(upperdata) or \"unknown\"  # use the get() method to search all dictionaries at once\n",
    "    upperResult = result.upper()\n",
    "    print(data+\" = \"+upperResult)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The two recommended solutions (pick one) at this point are to:\n",
    " - Email your code files to yourself\n",
    " - Copy your code files to a pen drive\n",
    " - Save your work on your own computer if you are using Anaconda ( or equiv)\n",
    "\n",
    "The class will be using canvas to submit the source code files created as part of this lab assignment.\n",
    "For this lab, you should upload the following programs as files:\n",
    " \n",
    " - this notebook file, complete with inspection introduction and results for coordMathSoln (4 points)\n",
    " - seqCleaner.py 9 points\n",
    " - fastqParse.py 9 points\n",
    " - coordinateMathSoln.py 9 points\n",
    " - converter.py 9 points\n",
    " - triad.py Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
